packages/permission-controller/src/PermissionController.ts

```diff
@@ -1993,6 +2036,90 @@ export class PermissionController<
     return approvedRequest as PermissionsRequest;
   }
 
+  /**
+   * Creates an arrray mapping of permitted and failure.
+   *
+   * @param permissions - The approved permissions.
+   * @returns The {@link SideEffects} object.
+   */
+  private getSideEffects(permissions: RequestedPermissions) {
+    return Object.keys(permissions).reduce<SideEffects>(
+      (sideEffectList, targetName) => {
+        const targetKey = this.getTargetKey(targetName);
+
+        if (targetKey) {
+          const specification = this.getPermissionSpecification(targetKey);
+
+          if (specification.sideEffect) {
+            sideEffectList.permittedHandlers.push(
+              specification.sideEffect.onPermitted,
+            );
+
+            if (specification.sideEffect.onFailure) {
+              sideEffectList.failureHandlers.push(
+                specification.sideEffect.onFailure,
+              );
+            }
+          }
+        }
+        return sideEffectList;
+      },
+      { permittedHandlers: [], failureHandlers: [] },
+    );
+  }
+
+  private async executeSideEffects(
+    sideEffects: SideEffects,
+    requestData: ApprovalRequestData,
+  ) {
+    const { permittedHandlers, failureHandlers } = sideEffects;
+    const params = {
+      requestData,
+      messagingSystem: this.messagingSystem,
+    };
+
+    const promiseResults = await Promise.allSettled(
+      permittedHandlers.map(async (permittedHandler) =>
+        permittedHandler(params),
+      ),
+    );
+
+    // lib.es2020.promise.d.ts does not export its types so we're using a simple type.
+    const rejectedHandlers = promiseResults.filter(
+      (promise) => promise.status === 'rejected',
+    ) as { status: 'rejected'; reason: Error }[];
+
+    if (rejectedHandlers.length > 0) {
+      try {
+        if (failureHandlers.length > 0) {
```

Comment:
We cannot merge the two if statements since it would mean adding an `else if` statemenent for all the things done ing there is a rejectedHandler but no failureHandlers

Reply:
I moved the try catch block under the second if statement though

Reply:
Ah, yeah. This is hard to read 😅